src/libraries/CalldataDecoder.sol

```diff
@@ -3,11 +3,93 @@ pragma solidity ^0.8.23;
 
 /// @title CalldataDecoder
 library CalldataDecoder {
+    using CalldataDecoder for bytes;
+    /// @notice equivalent to SliceOutOfBounds.selector, stored in least-significant bits
+
+    /// @notice mask used for offsets and lengths to ensure no overflow
+    /// @dev no sane abi encoding will pass in an offset or length greater than type(uint32).max
+    ///      (note that this does deviate from standard solidity behavior and offsets/lengths will
+    ///      be interpreted as mod type(uint32).max which will only impact malicious/buggy callers)
+    uint256 constant OFFSET_OR_LENGTH_MASK = 0xffffffff;
+    uint256 constant OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN = 0xffffffe0;
+
+    /// error SliceOutOfBounds();
+    uint256 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
```

Comment:
rm this we add it in the other PR when we do the length check in remove selector